File: source/viewdemo/toutiaodemo/src/main/java/com/zero/toutiaodemo/view/ColorChangeTextView2.java
package com.zero.toutiaodemo.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatTextView;

public class ColorChangeTextView2 extends AppCompatTextView {
    private static final String TAG = "Zero";

    private String mText = "享学课堂";//成员变量

    private float mPercent = 0.0f;

    public float getPercent() {
        return mPercent;
    }

    public void setPercent(float mPercent) {
        this.mPercent = mPercent;
        invalidate();
    }

    public ColorChangeTextView2(Context context) {
        this(context, null);
    }

    public ColorChangeTextView2(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ColorChangeTextView2(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init();
        initAttr(context, attrs);
    }

    private void init() {

    }

    private void initAttr(final Context context, @Nullable final AttributeSet attrs) {

    }

//    @Override
//    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
//        super.onMeasure(widthMeasureSpec, heightMeasureSpec);
//    }

    @Override
    protected void onDraw(final Canvas canvas) {
        super.onDraw(canvas);
        //绘制文字
        //决定显示什么(形状) canvas 画布
        //决定画的样式风格(大小，颜色) paint 画笔
        Paint paint = new Paint();
        paint.setTextSize(80);
        float baseline = 100;
        float x = getWidth()/2;
        canvas.drawText(mText,x,baseline,paint);
        //绘制辅助线
        drawCenterLineX(canvas);
        drawCenterLineY(canvas);
        //绘制到中间
        paint.setTextAlign(Paint.Align.CENTER);//第一种方法
        canvas.drawText(mText,x,baseline+paint.getFontSpacing(),paint);
        //绘制到右边
        paint.setTextAlign(Paint.Align.RIGHT);
        canvas.drawText(mText,x,baseline+paint.getFontSpacing()*2,paint);

        //第二种文字绘制到x中心
        drawCenterText(canvas);
        drawCenterText2(canvas);

    }

    //第二种文字绘制到x中心，第一层
    private void drawCenterText(final Canvas canvas){
        Paint paint = new Paint();
        paint.setStyle(Paint.Style.FILL);
        paint.setTextSize(80);
        paint.setColor(Color.BLACK);
        paint.setAntiAlias(true);
        //测量文字
        canvas.save();
        Paint.FontMetrics fontMetrics = paint.getFontMetrics();
        float height = fontMetrics.descent - fontMetrics.ascent;
        float baseline = getHeight()/2  - (fontMetrics.descent +fontMetrics.ascent)/2;
        float width = paint.measureText(mText);
        float left = getWidth()/2 - width/2;
        canvas.drawText(mText,left,baseline,paint);
        canvas.restore();
    }

    //第二层
    private void drawCenterText2(final Canvas canvas){
        Paint paint = new Paint();
        paint.setStyle(Paint.Style.FILL);
        paint.setTextSize(80);
        paint.setColor(Color.GREEN);
        paint.setAntiAlias(true);
        //测量文字
        canvas.save();
        Paint.FontMetrics fontMetrics = paint.getFontMetrics();
        float baseline = getHeight()/2  - (fontMetrics.descent +fontMetrics.ascent)/2;
        float width = paint.measureText(mText);
        float left = getWidth()/2 - width/2;

        float right = left + width*mPercent;
        Rect rect = new Rect((int)left,0,(int)right,getHeight());
        canvas.clipRect(rect);
        canvas.drawText(mText,left,baseline,paint);
        canvas.restore();
    }


    private void drawCenterLineX(final Canvas canvas){
        Paint paint = new Paint();
        paint.setStyle(Paint.Style.FILL);
        paint.setColor(Color.RED);
        paint.setStrokeWidth(3);
        canvas.drawLine(getWidth()/2,0,getWidth()/2,getHeight(),paint);
    }

    private void drawCenterLineY(final Canvas canvas){
        Paint paint = new Paint();
        paint.setStyle(Paint.Style.FILL);
        paint.setColor(Color.RED);
        paint.setStrokeWidth(3);
        canvas.drawLine(0,getHeight()/2,getWidth(),getHeight()/2,paint);
    }


}

File: source/viewdemo/toutiaodemo/src/main/java/com/zero/toutiaodemo/kotlin/ColorChangeTextView2.kt
package com.zero.toutiaodemo.kotlin

import android.content.Context
import android.content.res.TypedArray
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.util.AttributeSet
import android.util.Log
import android.view.View
import com.zero.toutiaodemo.R

class ColorChangeTextView2
@JvmOverloads
constructor(context: Context, attrs: AttributeSet, defStyle: Int = 0, defStyleRes: Int = 0)
    : View(context, attrs, defStyle, defStyleRes) {

    private var mPaint = Paint()//绘制文字
    private var mLinePaint = Paint()//辅助框

    private var mTextBound = Rect()
    private var mTextWidth = 0
    private var mTextHeight = 0

    private var mTextStartX = 0

    companion object{
        const val TAG = "Zero"

        const val DIRECTION_LEFT = 0
        const val DIRECTION_RIGHT = 1
        const val DIRECTION_TOP = 2
        const val DIRECTION_BOTTOM = 3
    }


    var text: String = "享学课堂"//属性  = java成员变量 + get set
        set(value) {
            field = value
            requestLayout()
            invalidate()
        }

    var textSize: Float = dp2px(30f)
        set(value) {
            field = value
            mPaint.textSize = value
            requestLayout()
            invalidate()
        }

    var textColor: Int = Color.BLACK
        set(value) {
            field = value
            invalidate()
        }
    var textColorChange: Int = Color.RED
        set(value) {
            field = value
            invalidate()
        }

    private var progress: Float = 0f
        set(value) {
            field = value
            invalidate()
        }

    var direction: Int = DIRECTION_LEFT

    init {
        val ta: TypedArray = context.obtainStyledAttributes(attrs, R.styleable.ColorChangeTextView1)
        text = ta.getString(R.styleable.ColorChangeTextView1_text) as String
        textSize = ta.getDimensionPixelSize(R.styleable.ColorChangeTextView1_text_size, textSize.toInt()).toFloat()
        textColor = ta.getColor(
                R.styleable.ColorChangeTextView1_text_color, textColor)
        textColorChange = ta.getColor(
                R.styleable.ColorChangeTextView1_text_color_change, textColorChange)
        progress = ta.getFloat(R.styleable.ColorChangeTextView1_progress, 0f)
        direction = ta
                .getInt(R.styleable.ColorChangeTextView1_direction, direction)
        ta.recycle()

        mPaint.textSize = textSize
    }

    init {
        mLinePaint.isAntiAlias = true
        mLinePaint.strokeWidth = sp2px(3f)
        mLinePaint.color = Color.GREEN
    }




    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
//        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        //确定自身的宽高
        //1. 测量出文字的大小
        measureText()
        //2. 确定自身的大小
        val width = measureWidth(widthMeasureSpec)
        val height = measureHeight(heightMeasureSpec)
        //3. 保存测量的尺寸 setMeasuredDimension()
        setMeasuredDimension(width,height)
        mTextStartX = measuredWidth/2 - mTextWidth/2

    }

    private fun measureWidth(measureSpec: Int): Int{
        var result = 0

        val mode =  MeasureSpec.getMode(measureSpec)//父view的mode
        val size = MeasureSpec.getSize(measureSpec)//父view的size

        when(mode){
            MeasureSpec.EXACTLY ->{
                result = size
            }
            MeasureSpec.AT_MOST,MeasureSpec.UNSPECIFIED ->{
                result = (mTextWidth +.5f).toInt() + paddingLeft + paddingRight
            }
        }

        result = if(mode == MeasureSpec.AT_MOST) result.coerceAtMost(size) else result
        return result
    }

    private fun measureHeight(measureSpec: Int): Int {
        val mode = MeasureSpec.getMode(measureSpec)
        val size = MeasureSpec.getSize(measureSpec)
        var result = 0
        when (mode) {
            MeasureSpec.EXACTLY -> {
                Log.i(TAG, "measureWidth: EXACTLY")
                result = size
            }
            MeasureSpec.AT_MOST, MeasureSpec.UNSPECIFIED -> result = (mTextBound.height() + .5f).toInt() + paddingTop + paddingBottom
        }
        //如果是AT_MOST,不能超过父布局的尺寸
        result = if (mode == MeasureSpec.AT_MOST) result.coerceAtMost(size) else result
        return result
    }

    private fun measureText(){
        mPaint.getTextBounds(text,0,text.length,mTextBound)
        mTextWidth = (mPaint.measureText(text) + .5f).toInt()

//        val fontMetrics = Paint.FontMetrics()
//        mPaint.getFontMetrics(fontMetrics)
//
//        mTextHeight = (fontMetrics.descent - fontMetrics.ascent + .5f).toInt()

    }


    override fun onDraw(canvas: Canvas?) {
        //可空对象 ？
        //不可空

        when(direction){
            DIRECTION_LEFT ->{
                canvas?.apply {
                    //绘制颜色改变的层
                    save()
                    mPaint.color = textColorChange
                    clipRect(mTextStartX,0,(mTextStartX + progress * mTextWidth).toInt(),measuredHeight)
                    drawText(text,mTextStartX.toFloat(),measuredHeight/2 - (mPaint.descent()/2 + mPaint.ascent()/2)
                            ,mPaint)
                    restore()

                    //绘制底色层
                    save()
                    mPaint.color = textColor
                    clipRect((mTextStartX + progress * mTextWidth).toInt(),0,mTextStartX + mTextWidth,measuredHeight)
                    drawText(text,mTextStartX.toFloat(),measuredHeight/2 - (mPaint.descent()/2 + mPaint.ascent()/2)
                            ,mPaint)
                    restore()

                }


            }
            DIRECTION_RIGHT ->{

            }
            DIRECTION_TOP ->{

            }
            DIRECTION_BOTTOM ->{

            }
        }
    }

}